### PR TITLE
feat(optimizer): refine the conversion from Join to Dynamic Filter

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

This PR refines the conversion from Join to Dynamic Filter.
- support all scalar inner sides
- ensure the output indices only from the outer side
- fix output indices column mapping

Also add some planner tests.

Note that some patterns can not be planned now due to pk derivation of hash join. Check the issue below.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
- #5764

### CODE REVIEW

File: src/frontend/planner_test/tests/testdata/dynamic_filter.yaml [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+              └─StreamHashAgg { group_key: [Vnode(t2._row_id)], aggs: [count, max(t2.v2)] }
+                └─StreamProject { exprs: [t2.v2, t2._row_id, Vnode(t2._row_id)] }
+                  └─StreamTableScan { table: t2, columns: [t2.v2, t2._row_id], pk: [t2._row_id], dist: UpstreamHashShard(t2._row_id) }
+- name: |
+    With Top-1 on inner side
+    TODO: currently not possible due to https://github.com/risingwavelabs/risingwave/issues/5764
+  before:
+  - create_tables
+  sql: |
+    with max_v2 as (select v2 max from t2 order by v2 desc limit 1) select v1 from t1, max_v2 where v1 > max;
```
Comment (chenzl25) [inline src/frontend/planner_test/tests/testdata/dynamic_filter.yaml:33]: Interesting.

**Review by chenzl25 — APPROVED** [review]
LGTM

File: src/frontend/planner_test/tests/testdata/dynamic_filter.yaml (hunk 1 above)
Comment (chenzl25) [reply src/frontend/planner_test/tests/testdata/dynamic_filter.yaml:33]: Some off-topic idea. Maybe we can support more range conditions than simple `>`, `>=`, `<`, `<=`.
e.g. `between`
```sql
 with max_v2 as (select max(v2) max from t2) select v1 from t1, max_v2 where v1 between max - 100 and max + 100;
```

**Review by st1page — APPROVED** [review]
LGTM, btw the `LogicalJoin::to_stream` is too long... It should be split into multiple functions. Maybe I will do it later

File: src/frontend/planner_test/tests/testdata/dynamic_filter.yaml (hunk 1 above)
Comment (BugenZhao) [reply src/frontend/planner_test/tests/testdata/dynamic_filter.yaml:33]: Exactly, we can support arbitrary range in the future.

**Pull request merged**